[BUG] Code doesn't compile with `val` in some cases

**Describe the bug**
Code with `val` don't compile with additional brackets. If change `val` to `final var` it compiles without errors

**To Reproduce**
```
private void df() {
        val typesForTesting = List.of("B", "R", "AR", "MB");
        val blaSetting = "bla-id.%s.%s.enabled";
        val someConstant = "GHJKJHG";
// works
        setSet("ahaha", "true");
// works
        setSet(("ahaha"), "true");
// works
        typesForTesting
                .forEach(type -> setSet("bla", "true"));
// works
        typesForTesting
                .forEach(type -> setSet(("bla"), "true"));
// works
        typesForTesting
                .forEach(type -> setSet(String.format(blaSetting, type, someConstant), "true"));

// doesn't work
        typesForTesting
                .forEach(type -> setSet((String.format(blaSetting, type, someConstant)), "true"));
    }

    private void setSet(String a, String b) {
    }
```

Stack trace
```
Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project test-cases: Compilation failure
An exception has occurred in the compiler (11.0.1). Please file a bug against the Java compiler via the Java bug reporting page (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com) for duplicates. Include your program and the following diagnostic in your report. Thank you.
java.lang.AssertionError: Incorrect number of arguments; expected 2, found 0
	at jdk.compiler/com.sun.tools.javac.util.Assert.error(Assert.java:162)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.visitApply(TransTypes.java:636)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1634)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:58)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.translate(TransTypes.java:454)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.visitLambda(TransTypes.java:540)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCLambda.accept(JCTree.java:1807)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:58)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.translate(TransTypes.java:454)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.translateArgs(TransTypes.java:212)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.visitApply(TransTypes.java:639)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1634)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:58)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.translate(TransTypes.java:454)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.visitExec(TransTypes.java:599)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1452)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:58)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:70)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.visitBlock(TreeTranslator.java:167)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1020)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:58)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.translate(TransTypes.java:454)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.visitMethodDef(TransTypes.java:489)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:866)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:58)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:70)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.visitClassDef(TreeTranslator.java:139)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.translateClass(TransTypes.java:918)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.visitClassDef(TransTypes.java:475)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:774)
	at jdk.compiler/com.sun.tools.javac.tree.TreeTranslator.translate(TreeTranslator.java:58)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.translate(TransTypes.java:454)
	at jdk.compiler/com.sun.tools.javac.comp.TransTypes.translateTopLevelClass(TransTypes.java:942)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.desugar(JavaCompiler.java:1553)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.desugar(JavaCompiler.java:1432)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.desugar(JavaCompiler.java:1516)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.desugar(JavaCompiler.java:1432)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)

```

**Expected behavior**
Code should be compiled without error

**Version info (please complete the following information):**
 - Lombok version 1.18.12
 - Platform MacOS 10.14.6  
java --version
openjdk 11.0.1 2018-10-16
OpenJDK Runtime Environment 18.9 (build 11.0.1+13)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.1+13, mixed mode)

--- a/src/core/lombok/javac/JavacResolution.java
+++ b/src/core/lombok/javac/JavacResolution.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2011-2020 The Project Lombok Authors.
+ * Copyright (C) 2011-2021 The Project Lombok Authors.
  * 
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
@@ -29,6 +29,7 @@ import java.lang.reflect.Method;
 import java.util.ArrayDeque;
 import java.util.Collection;
 import java.util.Iterator;
+import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.NoSuchElementException;
 
@@ -45,6 +46,7 @@ import com.sun.tools.javac.code.Type.ClassType;
 import com.sun.tools.javac.code.Type.TypeVar;
 import com.sun.tools.javac.code.Type.WildcardType;
 import com.sun.tools.javac.code.Types;
+import com.sun.tools.javac.comp.ArgumentAttr;
 import com.sun.tools.javac.comp.Attr;
 import com.sun.tools.javac.comp.AttrContext;
 import com.sun.tools.javac.comp.Enter;
@@ -68,6 +70,7 @@ import lombok.core.debug.AssertionLogger;
 import lombok.permit.Permit;
 
 public class JavacResolution {
+	private final Context context;
 	private final Attr attr;
 	private final CompilerMessageSuppressor messageSuppressor;
 	
@@ -82,6 +85,7 @@ public class JavacResolution {
 	}
 	
 	public JavacResolution(Context context) {
+		this.context = context;
 		attr = Attr.instance(context);
 		messageSuppressor = new CompilerMessageSuppressor(context);
 	}
@@ -245,10 +249,19 @@ public class JavacResolution {
 		} catch (Throwable ignore) {
 			// This addresses issue #1553 which involves JDK9; if it doesn't exist, we probably don't need to set it.
 		}
-		if (tree instanceof JCBlock) attr.attribStat(tree, env);
-		else if (tree instanceof JCMethodDecl) attr.attribStat(((JCMethodDecl) tree).body, env);
-		else if (tree instanceof JCVariableDecl) attr.attribStat(tree, env);
-		else throw new IllegalStateException("Called with something that isn't a block, method decl, or variable decl");
+		
+		Map<?,?> cache = null;
+		try {
+			cache = ArgumentAttrReflect.enableTempCache(context);
+			
+			if (tree instanceof JCBlock) attr.attribStat(tree, env);
+			else if (tree instanceof JCMethodDecl) attr.attribStat(((JCMethodDecl) tree).body, env);
+			else if (tree instanceof JCVariableDecl) attr.attribStat(tree, env);
+			else throw new IllegalStateException("Called with something that isn't a block, method decl, or variable decl");
+		} finally {
+			ArgumentAttrReflect.restoreCache(cache, context);
+		}
+		
 	}
 	
 	public static class TypeNotConvertibleException extends Exception {
@@ -280,6 +293,43 @@ public class JavacResolution {
 		
 		public static Type Types_upperBound(Types types, Type type) {
 			return (Type) Permit.invokeSneaky(initError, UPPER_BOUND, types, type);
+		}
+	}
+	
+	/**
+	 * ArgumentAttr was added in Java 9 and caches some method arguments. Lombok should cleanup its changes after resolution.
+	 */
+	private static class ArgumentAttrReflect {
+		private static Field ARGUMENT_TYPE_CACHE;
+		
+		static {
+			if (Javac.getJavaCompilerVersion() >= 9) {
+				try {
+					ARGUMENT_TYPE_CACHE = Permit.getField(ArgumentAttr.class, "argumentTypeCache");
+				} catch (Exception ignore) {}
+			}
+		}
+		
+		public static Map<?, ?> enableTempCache(Context context) {
+			if (ARGUMENT_TYPE_CACHE == null) return null;
+			
+			ArgumentAttr argumentAttr = ArgumentAttr.instance(context);
+			try {
+				Map<?, ?> cache = (Map<?, ?>) Permit.get(ARGUMENT_TYPE_CACHE, argumentAttr);
+				Permit.set(ARGUMENT_TYPE_CACHE, argumentAttr, new LinkedHashMap<Object, Object>(cache));
+				return cache;
+			} catch (Exception ignore) { }
+			
+			return null;
+		}
+		
+		public static void restoreCache(Map<?, ?> cache, Context context) {
+			if (ARGUMENT_TYPE_CACHE == null) return;
+			
+			ArgumentAttr argumentAttr = ArgumentAttr.instance(context);
+			try {
+				Permit.set(ARGUMENT_TYPE_CACHE, argumentAttr, cache);
+			} catch (Exception ignore) { }
 		}
 	}
 	

--- a/src/stubs/com/sun/tools/javac/comp/ArgumentAttr.java
+++ b/src/stubs/com/sun/tools/javac/comp/ArgumentAttr.java
@@ -1,0 +1,12 @@
+/*
+ * These are stub versions of various bits of javac-internal API (for various different versions of javac). Lombok is compiled against these.
+ */
+package com.sun.tools.javac.comp;
+
+import com.sun.tools.javac.tree.JCTree;
+import com.sun.tools.javac.util.Context;
+
+// JDK9+
+public class ArgumentAttr extends JCTree.Visitor {
+	public static ArgumentAttr instance(Context context) { return null; }
+}

